TELCODOCS-1707: Preparing generate seed image (#77308)
openshift/openshift-docs: modules/cnf-image-based-upgrade-generate-seed-image.adoc

// Module included in the following assemblies:
// * edge_computing/image-based-upgrade/cnf-preparing-for-image-based-upgrade.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image with the `SeedGenerator` CR. The Operator checks for required system configurations, performs any necessary system cleanup before generating the seed image, and launches the image generation. The seed image generation includes the following tasks:

> * Stopping cluster operators

Reviewer: ```suggestion
* Stopping cluster Operators
```

* Preparing the seed image configuration
* Generating and pushing the seed image to the image repository specified in the `SeedGenerator` CR

> * Restoring cluster operators

Reviewer: ```suggestion
* Restoring cluster Operators
```

* Expiring seed cluster certificates
* Generating new certificates for the seed cluster
* Restoring and updating the `SeedGenerator` CR on the seed cluster

.Prerequisites

* Configure a shared container directory on the seed cluster.
* Install the OADP Operator and the {lcao} on the seed cluster.

.Procedure

. Detach the cluster from the hub to delete any cluster-specific resources from the seed cluster that must not be in the seed image:

.. If you are using {rh-rhacm}, manually detach the seed cluster by running the following command:
+
[source,terminal]
----
$ oc delete managedcluster sno-worker-example
----

> ... Wait until the `ManagedCluster` CR is removed. Once the CR is removed, create the proper `SeedGenerator` CR. The {lcao} cleans up the {rh-rhacm} artifacts.

Reviewer: ```suggestion
... Wait until the `ManagedCluster` CR is removed. After the CR is removed, create the proper `SeedGenerator` CR. The {lcao} cleans up the {rh-rhacm} artifacts.
```

.. If you are using GitOps ZTP, detach your cluster by removing the seed cluster's `SiteConfig` CR from the `kustomization.yaml`:

... Remove your seed cluster's `SiteConfig` CR from the `kustomization.yaml`.
+
[source,yaml]
----
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

generators:
#- example-seed-sno1.yaml
- example-target-sno2.yaml
- example-target-sno3.yaml
----

... Commit the `kustomization.yaml` changes in your Git repository and push the changes.

+
The ArgoCD pipeline detects the changes and removes the managed cluster.

. Create the `Secret`.

.. Create the authentication file by running the following command:
+
--
.Authentication file
[source,terminal]

> ----

Reviewer: This appears to copy correct. Would you happen to know offhand if copying and pasting this command in your terminal will run all three steps correct? I'm assuming that it will. If it does not, you might consider having three separate code blocks for this content.

Reviewer 2: I just ran this in my terminal, of course podman couldn't log into the nonexistent registry but all of them executed, so it works :)

$ MY_USER=myuserid
$ AUTHFILE=/tmp/my-auth.json
$ podman login --authfile ${AUTHFILE} -u ${MY_USER} quay.io/${MY_USER}
----

[source,terminal]
----
$ base64 -w 0 ${AUTHFILE} ; echo
----
--

.. Copy the output into the `seedAuth` field in the `Secret` YAML file named `seedgen` in the `openshift-lifecycle-agent` namespace.
+
--
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: seedgen <1>
  namespace: openshift-lifecycle-agent
type: Opaque
data:
  seedAuth: <encoded_AUTHFILE> <2>
----
<1> The `Secret` resource must have the `name: seedgen` and `namespace: openshift-lifecycle-agent` fields.
<2> Specifies a base64-encoded authfile for write-access to the registry for pushing the generated seed images.
--

> .. Apply the `Secret`.

Reviewer: ```suggestion
.. Apply the `Secret` by running the following command:
```

+
[source,terminal]
----
$ oc apply -f secretseedgenerator.yaml
----

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image:<tag>`. It is recommended to use the `<seed_cluster_name>:<ocp_version>` format.
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster reboots and loses API capabilities while the {lcao} generates the seed image.
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, then it starts the image generation.
====

> If you want to generate further seed images, you must provision a new seed cluster with the version you want to generate a seed image from.

Reviewer: Is "further" correct? It might just be context that I do not understand. "more"? 

```suggestion
If you want to generate further seed images, you must provision a new seed cluster with the version that you want to generate a seed image from.
```

Reviewer 2: Yeah, more should work here too

.Verification

> . Once the cluster recovers and it is available, you can check the status of the `SeedGenerator` CR:

Reviewer: I'm not really a stickler for this rule, but I always note it: 

```suggestion
. After the cluster recovers and it is available, you can check the status of the `SeedGenerator` CR by running the following command:
```
